caput/memh5.py

```diff
@@ -2400,3 +2566,207 @@ def bytes_to_unicode(s):
+
+            items.append((name, newtype))
+
+        return items
+
+    # For scalar types the conversion is easy
+    if not dt.names:
+        return np.dtype(_conv(dt.str))
+    # For compound types we need to iterate through
+    else:
+        return np.dtype(_iter_conv(dt.descr))
+
+
+def has_kind(dt, kind):
+    """Test if a numpy datatype has any fields of a specified type.
+
+    Parameters
+    ----------
+    dt : np.dtype
+        Data type to convert.
+    kind : str
+        Numpy type code character. e.g. "S" for bytestring and "U" for unicode.
+
+    Returns
+    ------
+    has_kind : bool
+        True if it contains the requested kind.
+    """
+
+    # For scalar types the conversion is easy
+    if not dt.names:
+        return dt.kind == kind
+
+    # For compound types we must recurse over the full compound type structure
+    def _iter_conv(x):
+
+        for item in x:
+            type_ = item[1]
+
+            # Recursively convert the type
+            if isinstance(type_, list) and _iter_conv(type_):
+                return True
+            elif type_[1] == kind:
+                return True
+
+        return False
+
+    return _iter_conv(dt.descr)
+
+
+def has_unicode(dt):
+    """Test if data type contains any unicode fields.
+
+    See `has_kind`.
+    """
+    return has_kind(dt, "U")
+
+
+def has_bytestring(dt):
+    """Test if data type contains any unicode fields.
+
+    See `has_kind`.
+    """
+    return has_kind(dt, "S")
+
+
+def ensure_bytestring(arr):
+    """If needed convert the array to contain bytestrings not unicode.
+
+    Parameters
+    ----------
+    arr : np.ndarray
+        Input array.
+
+    Returns
+    -------
+    arr_conv : np.ndarray
+        The converted array. If no conversion was required, just returns `arr`.
+    """
+    if has_unicode(arr.dtype):
+        return arr.astype(dtype_to_bytestring(arr.dtype))
+    else:
+        return arr
+
+
+def ensure_unicode(arr):
+    """If needed convert the array to contain unicode strings not bytestrings.
+
+    Parameters
+    ----------
+    arr : np.ndarray
+        Input array.
+
+    Returns
+    -------
+    arr_conv : np.ndarray
+        The converted array. If no conversion was required, just returns `arr`.
+    """
+    if has_bytestring(arr.dtype):
+        return arr.astype(dtype_to_unicode(arr.dtype))
+    else:
+        return arr
+
+
+def check_unicode(dset):
+    """Test if dataset contains unicode so we can raise an appropriate error.
+
+    If there is no unicode, return the data from the array.
+
+    Parameters
+    ----------
+    dset : MemDataset
+        Dataset to check.
+
+    Returns
+    -------
+    dset :
+        The converted array. If no conversion was required, just returns `arr`.
+    """
+    if has_unicode(dset.dtype):
+        raise TypeError(
+            'Can not write dataset "%s" of unicode type into HDF5.' % dset.name
+        )
+    else:
```

Comment:
I guess you can argue about this, but I find that there should really not be an `else` here.